python/ray/data/impl/pipeline_executor.py

```diff
@@ -49,51 +50,53 @@ def __iter__(self):
         return self
 
     def __next__(self):
-        output = None
-        start = time.perf_counter()
-
-        while output is None:
-            if all(s is None for s in self._stages):
-                raise StopIteration
-
-            # Wait for any completed stages.
-            pending = [s for s in self._stages if s is not None]
-            ready, _ = ray.wait(pending, timeout=0.1, num_returns=len(pending))
-
-            # Bubble elements down the pipeline as they become ready.
-            for i in range(len(self._stages))[::-1]:
-                is_last = i + 1 >= len(self._stages)
-                next_slot_free = is_last or self._stages[i + 1] is None
-                if not next_slot_free:
-                    continue
-
-                slot_ready = self._stages[i] in ready
-                if not slot_ready:
-                    continue
-
-                # Bubble.
-                result = ray.get(self._stages[i])
-                if self._bars:
-                    self._bars[i].update(1)
-                self._stages[i] = None
-                if is_last:
-                    output = result
-                else:
-                    fn = self._pipeline._stages[i]
-                    self._stages[i + 1] = pipeline_stage.remote(
-                        lambda: fn(result), DatasetContext.get_current())
-
-            # Pull a new element for the initial slot if possible.
-            if self._stages[0] is None:
-                try:
-                    self._stages[0] = pipeline_stage.remote(
-                        next(self._iter), DatasetContext.get_current())
-                except StopIteration:
-                    pass
-
-        self._pipeline._stats.wait_time_s.append(time.perf_counter() - start)
-        self._pipeline._stats.add(output._stats)
-        return output
+        with concurrent.futures.ThreadPoolExecutor() as pool:
```

Comment:
Note that this will create and destroy a threadpool on every iteration of the executor, which should be equal to the length of the pipeline (number of stages). What do you think about doing something like tying the lifetime of the threadpool to the `PipelineExecutor`?

```python
class PipelineExecutor:
    def __init__(...):
        # ...
        self._pool = concurrent.futures.ThreadPoolExecutor()
        self._stages[0] = self._pool.submit(lambda: pipeline_stage(
                            next(self._iter), DatasetContext.get_current()))

    def __next__(self):
        # ...
        while output is None:
            if all(s is None for s in self._stages):
                self._pool.shutdown()
                raise StopIteration
```